django/db/models/expressions.py

```diff
@@ -817,7 +817,7 @@ class When(Expression):
     def __init__(self, condition=None, then=None, **lookups):
         if lookups and condition is None:
             condition, lookups = Q(**lookups), None
-        if condition is None or not isinstance(condition, Q) or lookups:
+        if condition is None or not getattr(condition, 'conditional', False) or lookups:
```

Comment:
We could also allow `WhereNode` type but that would require a conditional import because of an import cycle. I figured duck typing would work just fine here given the fact `WhereNode` is really just a _resolved_ `Q`.